### DISCUSSION TITLE
CGAL geometry kernel primitives

### DISCUSSION THREAD

**Discussion by petrasvestartas** [CGAL/cgal#6006] [Q&A] [answered]
Hi,

I have very naive and simple question, but it pops out from time to time.

How do you normally copy simple geometry kernels such point, line, segement, direction, plane?

Is it enough to write:
Point a = b;
Or I always must use constructor:
Point a(b);

**Comment by afabri** [comment]
`a = b;`  is fine.

**Reply by petrasvestartas** [reply]
Thank you

**Comment by lrineau** [comment, marked as answer]
You question is a C++ question, and it is not specific to CGAL. All you have to know is that geometry kernel primitives are class templates, and they have copy constructors.

## Copy initialization
```c++
Point a = b;
```
This is a [copy initialization](https://en.cppreference.com/w/cpp/language/copy_initialization). As `Point` is a class, and both `a` and `b` are of type `Point`, and `b` is not a prvalue, then the copy constructor of `Point` is called to initialize `a` with a copy of `b`.

## Direct initialization
```c++
Point a(b);
```
This is a [direct initilization](https://en.cppreference.com/w/cpp/language/direct_initialization). `Point` is a class and both `a` and `b` are of type `Point`, and `b` is not a prvalue, then the copy constructor of `Point` is called to initialize `a` with a copy of `b`.

## Conclusion
```c++
Point a = b;
Point aa(b);
```
For CGAL geometric objects, those two ways of initialization are completely equivalent.

## Side-note
```c++
Point a;
a = b
```
That is completely different: that is a [value initialization](https://en.cppreference.com/w/cpp/language/value_initialization) (that class the default constructor) followed by a [assignment](https://en.cppreference.com/w/cpp/language/operator_assignment#Builtin_direct_assignment) (that calls the copy assignment operator of the class).
